include/deal.II/base/mpi_compute_index_owner_internal.h

```diff
@@ -119,6 +119,10 @@ namespace Utilities
             this->partition(owned_indices, comm);
 
 #ifdef DEAL_II_WITH_MPI
+
+            static CollectiveMutex      mutex;
+            CollectiveMutex::ScopedLock lock(mutex, comm);
```

Comment:
Can't you scope this better? I think it only needs to cover steps 3 and 4 below.

Reply:
like this?